chiron/mcmc.py

```diff
@@ -112,103 +170,802 @@ def run(
+        calculate_current_reduced_potential : bool, optional
+            Whether to calculate the current reduced potential. Default is True.
 
         Returns
         -------
-        float
-            Acceptance probability as a float.
+        sampler_state : SamplerState
+            The updated sampler state; if a move is rejected this will be unchanged.
+            Note, if the proposed move is rejected, the current PRNG key will be updated to ensure
+            that we are using a different random number for the next iteration.
+        thermodynamic_state : ThermodynamicState
+            The updated thermodynamic state; if a move is rejected this will be unchanged.
+            Note, many MC moves will not modify the thermodynamic state regardless of acceptance of the move.
+        nbr_list: PairsBase, optional
+            The updated neighbor/pair list. If a nbr_list is not set, this will be None.
+            If the move is rejected, this will correspond to the neighbor
+
+        """
+
+        # if this is the first time we are calling this function during this iteration
+        # we will need to calculate the reduced potential for the current state
+        # this is toggled by the calculate_current_reduced_potential flag
+        # otherwise, we can use the one that was saved from the last step, for efficiency
+        if calculate_current_reduced_potential:
+            current_reduced_pot = current_thermodynamic_state.get_reduced_potential(
+                current_sampler_state, current_nbr_list
+            )
+            # save the current_reduced_pot so we don't have to recalculate
+            # it on the next iteration if the move is rejected
+            self._current_reduced_pot = current_reduced_pot
+        else:
+            current_reduced_pot = self._current_reduced_pot
+
+        # propose a new state and calculate the log proposal ratio
+        # this will be specific to the type of move
+        # in addition to the sampler_state, this will require/return the thermodynamic state
+        # for systems that e.g., make changes to particle identity.
+        # For efficiency, we will also return a copy of the nbr_list associated with the proposed state
+        # because if the move is rejected, we can move back the original state without having to rebuild the nbr_list
+        # if it were modified due to the proposed state.
+        (
+            proposed_sampler_state,
+            proposed_thermodynamic_state,
+            proposed_reduced_pot,
+            log_proposal_ratio,
+            proposed_nbr_list,
+        ) = self._propose(
+            current_sampler_state,
+            current_thermodynamic_state,
+            current_reduced_pot,
+            current_nbr_list,
+        )
+
+        if jnp.isnan(proposed_reduced_pot):
+            decision = False
+        else:
+            # accept or reject the proposed state
+            decision = self._accept_or_reject(
```

Comment:
I don't like the idea of having a method like `self._accept_or_reject()` use some object fields (current and previous reduced potential energies) and some arguments (log proposal ratio, PRNG, and method---which is already stored as an object field) to perform the calculations. Especially thinking ahead to perhaps making the acceptance behavior a class, it would make more sense if we just forced everything to be an argument for a mathematics-based computation like the acceptance probability here. 

Even better woudl be to have it return the log probability of acceptance, since we could store that during `self._update_statistics()` in addition to the `decision`. The statistics of `log(Paccept)` are valuable in figuring out how bad of an acceptance problem we have, and can even be used to estimate very small acceptance ratios.

Reply:
I copied this text into issue #35 that discusses making acceptance criteria modular.